File: sdks/java/src/main/java/org/byteveda/taskito/proxies/Proxies.java
package org.byteveda.taskito.proxies;

import com.fasterxml.jackson.databind.ObjectMapper;
import com.fasterxml.jackson.databind.SerializationFeature;
import java.nio.charset.StandardCharsets;
import java.security.MessageDigest;
import java.time.Duration;
import java.util.Base64;
import java.util.LinkedHashMap;
import java.util.Map;
import javax.crypto.Mac;
import javax.crypto.spec.SecretKeySpec;
import org.byteveda.taskito.errors.ProxyException;

/**
 * Registry that deconstructs resources into signed {@link ProxyRef}s and
 * reconstructs them. Construct with an HMAC key (shared by producer and worker),
 * register a {@link ProxyHandler} per resource type, then
 * {@link #deconstruct(Object)} on the producer and {@link #reconstruct(ProxyRef)}
 * (or {@link #resolve(ProxyRef)}) on the worker.
 */
public final class Proxies {
    private static final String ALGORITHM = "HmacSHA256";

    private final Map<String, ProxyHandler<?>> handlers = new LinkedHashMap<>();
    private final byte[] key;
    private final ObjectMapper canonical =
            new ObjectMapper().configure(SerializationFeature.ORDER_MAP_ENTRIES_BY_KEYS, true);

    public Proxies(byte[] hmacKey) {
        this.key = hmacKey.clone();
    }

    /** Register a handler under its non-null, unique id; returns {@code this}. */
    public Proxies register(ProxyHandler<?> handler) {
        String id = handler.id();
        if (id == null) {
            throw new ProxyException("proxy handler id must not be null");
        }
        // Fail fast on a duplicate: silently overwriting would let a producer and
        // worker disagree on what a given ProxyRef's handler id means.
        if (handlers.putIfAbsent(id, handler) != null) {
            throw new ProxyException("proxy handler '" + id + "' is already registered");
        }
        return this;
    }

    /** Deconstruct {@code value} into a signed ref with no expiry or purpose. */
    public ProxyRef deconstruct(Object value) {
        return deconstruct(value, null, null);
    }

    /** Deconstruct {@code value} into a ref that expires after {@code ttl}. */
    public ProxyRef deconstruct(Object value, Duration ttl) {
        return deconstruct(value, ttl, null);
    }

    /**
     * Deconstruct {@code value} into a signed ref bound to {@code ttl} (nullable)
     * and {@code purpose} (nullable); throws if no handler accepts it.
     */
    @SuppressWarnings("unchecked")
    public ProxyRef deconstruct(Object value, Duration ttl, String purpose) {
        if (value == null) {
            throw new ProxyException("cannot deconstruct null");
        }
        Long expiresAtMs = ttl == null ? null : System.currentTimeMillis() + ttl.toMillis();
        for (ProxyHandler<?> handler : handlers.values()) {
            if (handler.handles(value)) {
                Map<String, Object> reference = ((ProxyHandler<Object>) handler).deconstruct(value);
                String signature = sign(handler.id(), reference, expiresAtMs, purpose);
                return new ProxyRef(handler.id(), reference, signature, expiresAtMs, purpose);
            }
        }
        throw new ProxyException("no proxy handler for " + value.getClass().getName());
    }

    /** Verify a ref's signature and expiry, then reconstruct the resource. */
    public Object reconstruct(ProxyRef ref) {
        return reconstruct(ref, null);
    }

    /**
     * Verify a ref's signature, expiry, and (when {@code expectedPurpose} is
     * non-null) its bound purpose, then reconstruct the resource.
     */
    public Object reconstruct(ProxyRef ref, String expectedPurpose) {
        ProxyHandler<Object> handler = handlerFor(ref.handler());
        verify(ref, expectedPurpose);
        return handler.reconstruct(ref.reference());
    }

    /**
     * A new {@link ProxySession}: deconstruct/reconstruct with per-instance
     * identity dedup and close-time cleanup. Confine a session to one thread.
     */
    public ProxySession session() {
        return new ProxySession(this);
    }

    /** Look up a handler by id, failing on an unknown id. */
    @SuppressWarnings("unchecked")
    ProxyHandler<Object> handlerFor(String handlerId) {
        ProxyHandler<Object> handler = (ProxyHandler<Object>) handlers.get(handlerId);
        if (handler == null) {
            throw new ProxyException("unknown proxy handler '" + handlerId + "'");
        }
        return handler;
    }

    /** Verify a ref's signature, expiry, and (when requested) bound purpose. */
    void verify(ProxyRef ref, String expectedPurpose) {
        byte[] expected = sign(ref.handler(), ref.reference(), ref.expiresAtMs(), ref.purpose())
                .getBytes(StandardCharsets.UTF_8);
        byte[] actual = (ref.signature() == null ? "" : ref.signature()).getBytes(StandardCharsets.UTF_8);
        if (!MessageDigest.isEqual(expected, actual)) {
            throw new ProxyException("proxy signature mismatch for handler '" + ref.handler() + "'");
        }
        if (ref.expiresAtMs() != null && System.currentTimeMillis() > ref.expiresAtMs()) {
            throw new ProxyException("proxy ref expired for handler '" + ref.handler() + "'");
        }
        if (expectedPurpose != null && !expectedPurpose.equals(ref.purpose())) {
            throw new ProxyException("proxy purpose mismatch for handler '" + ref.handler() + "'");
        }
    }

    /** {@link #reconstruct(ProxyRef)} cast to the caller's type. */
    @SuppressWarnings("unchecked")
    public <T> T resolve(ProxyRef ref) {
        return (T) reconstruct(ref);
    }

    /** {@link #reconstruct(ProxyRef, String)} cast to the caller's type. */
    @SuppressWarnings("unchecked")
    public <T> T resolve(ProxyRef ref, String expectedPurpose) {
        return (T) reconstruct(ref, expectedPurpose);
    }

    private String sign(String handlerId, Map<String, Object> reference, Long expiresAtMs, String purpose) {
        try {
            Mac mac = Mac.getInstance(ALGORITHM);
            mac.init(new SecretKeySpec(key, ALGORITHM));
            mac.update(handlerId.getBytes(StandardCharsets.UTF_8));
            mac.update((byte) '\n');
            mac.update(canonical.writeValueAsBytes(reference));
            mac.update((byte) '\n');
            mac.update(String.valueOf(expiresAtMs).getBytes(StandardCharsets.UTF_8));
            mac.update((byte) '\n');
            mac.update((purpose == null ? "" : purpose).getBytes(StandardCharsets.UTF_8));
            return Base64.getEncoder().encodeToString(mac.doFinal());
        } catch (Exception e) {
            throw new ProxyException("failed to sign proxy ref", e);
        }
    }
}

File: sdks/node/src/proxies/proxies.ts
import { createHmac, timingSafeEqual } from "node:crypto";
import { ProxyError } from "../errors";
import { canonicalJson } from "./canonical";
import { ProxySession } from "./proxy-session";
import type { ProxyHandler, ProxyRef } from "./types";

/**
 * Registry that deconstructs resources into signed {@link ProxyRef}s and
 * reconstructs them. Construct with an HMAC key (shared by producer and
 * worker), register a {@link ProxyHandler} per resource type, then
 * `deconstruct` on the producer and `reconstruct` (or `resolve`) on the
 * worker. Signature layout is the cross-SDK contract, so refs verify across
 * SDKs sharing the key.
 */
export class Proxies {
  private readonly handlers = new Map<string, ProxyHandler>();
  private readonly key: Buffer;

  constructor(hmacKey: Uint8Array) {
    if (hmacKey.length === 0) {
      throw new ProxyError("proxy HMAC key must not be empty");
    }
    this.key = Buffer.from(hmacKey);
  }

  /** Register a handler under its non-empty, unique id; returns `this`. */
  register<T>(handler: ProxyHandler<T>): this {
    if (!handler.id) {
      throw new ProxyError("proxy handler id must not be empty");
    }
    // Fail fast on a duplicate: silently overwriting would let a producer and
    // worker disagree on what a given ProxyRef's handler id means.
    if (this.handlers.has(handler.id)) {
      throw new ProxyError(`proxy handler "${handler.id}" is already registered`);
    }
    this.handlers.set(handler.id, handler as ProxyHandler);
    return this;
  }

  /**
   * Deconstruct `value` into a signed ref, optionally bound to a TTL and a
   * purpose; throws if no registered handler accepts it (registration order,
   * first match wins).
   */
  deconstruct(value: unknown, opts?: { ttlMs?: number; purpose?: string }): ProxyRef {
    if (value === null || value === undefined) {
      throw new ProxyError("cannot deconstruct null");
    }
    const expiresAtMs = opts?.ttlMs === undefined ? null : Date.now() + opts.ttlMs;
    const purpose = opts?.purpose ?? null;
    for (const handler of this.handlers.values()) {
      if (handler.handles(value)) {
        const reference = handler.deconstruct(value);
        const signature = this.sign(handler.id, reference, expiresAtMs, purpose);
        return { handler: handler.id, reference, signature, expiresAtMs, purpose };
      }
    }
    throw new ProxyError(`no proxy handler for value of type ${typeof value}`);
  }

  /**
   * Verify a ref's signature, expiry, and (when `expectedPurpose` is given)
   * its bound purpose, then reconstruct the resource.
   */
  reconstruct(ref: ProxyRef, expectedPurpose?: string): unknown {
    const handler = this.handlerFor(ref.handler);
    this.verifyRef(ref, expectedPurpose);
    return handler.reconstruct(ref.reference);
  }

  /** {@link Proxies.reconstruct} cast to the caller's type. */
  resolve<T>(ref: ProxyRef, expectedPurpose?: string): T {
    return this.reconstruct(ref, expectedPurpose) as T;
  }

  /**
   * Open a {@link ProxySession} over this registry — a unit-of-work wrapper
   * adding identity dedup on deconstruct, memoization on reconstruct, and a
   * LIFO cleanup lifecycle on close.
   */
  session(): ProxySession {
    return new ProxySession(this);
  }

  /** Look up the handler registered under `handlerId`. @internal */
  handlerFor(handlerId: string): ProxyHandler {
    const handler = this.handlers.get(handlerId);
    if (!handler) {
      throw new ProxyError(`unknown proxy handler "${handlerId}"`);
    }
    return handler;
  }

  /** Verify a ref's signature, expiry, and (optionally) purpose. @internal */
  verifyRef(ref: ProxyRef, expectedPurpose?: string): void {
    const expected = Buffer.from(
      this.sign(ref.handler, ref.reference, ref.expiresAtMs, ref.purpose),
      "utf8",
    );
    const actual = Buffer.from(ref.signature ?? "", "utf8");
    // timingSafeEqual throws on length mismatch — treat that as a plain mismatch.
    if (expected.length !== actual.length || !timingSafeEqual(expected, actual)) {
      throw new ProxyError(`proxy signature mismatch for handler "${ref.handler}"`);
    }
    if (ref.expiresAtMs != null && Date.now() > ref.expiresAtMs) {
      throw new ProxyError(`proxy ref expired for handler "${ref.handler}"`);
    }
    if (expectedPurpose !== undefined && (ref.purpose ?? null) !== expectedPurpose) {
      throw new ProxyError(`proxy purpose mismatch for handler "${ref.handler}"`);
    }
  }

  private sign(
    handlerId: string,
    reference: Record<string, unknown>,
    expiresAtMs: number | null | undefined,
    purpose: string | null | undefined,
  ): string {
    // Cross-SDK contract: an absent expiry signs as the literal string "null";
    // an absent purpose signs as the empty string.
    const message = [
      handlerId,
      canonicalJson(reference),
      expiresAtMs == null ? "null" : String(expiresAtMs),
      purpose ?? "",
    ].join("\n");
    return createHmac("sha256", this.key).update(message, "utf8").digest("base64");
  }
}
